### PR TITLE
Feature/enforce coding style

### PR DESCRIPTION
**Pull request by markmcdowall**
## Description

This brings in tests for pod coverage and perlcritic as part of the testing. The The config files for perlcritic and perltidy have been renamed to .perlcriticrc and .perltidyrc as expected by the relevant analysis tools.

For perlcritic, if it is run locally then it runs as brutal, however in the results errors that are of levels 3, 4 or 5 are coloured. On travis only those that are coloured (--harsh) are tested for.

For the pod coverage it will highlight public functions that have not been annotated. All public (and private) functions should get annotated.

## Benefits

This is to help improve the readability of the code as it is developed rather than requiring large refactoring efforts at a later stage.

## Possible Drawbacks

There will be an initial increase in the errors that are raised

## Testing

These are tests that will cause Travis to fail until the perlcritic and pod documentation has been added.

### CODE REVIEW

File: .perlcriticrc [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+perltidyrc = .perltidyrc
+
+# [Subroutines::ProhibitUnusedPrivateSubroutines]
+# allow = _init_db _init_config _guess_config
+
+[Subroutines::ProhibitUnusedPrivateSubroutines]
+skip_when_using = Moose
+
+[Subroutines::ProhibitManyArgs]
+max_arguments = 7
```
Comment (mkszuba) [inline .perlcriticrc]: Please add a newline at the end.

**Comment by markmcdowall** [comment]
@mkszuba Checking your code locally for the files that you have changed should suffice for the moment.

As for the legacy code this is the reason to make these changes and add perlcritic sooner rather than later given that this is a new repository structure. I have fixes already for many of the issues on a separate branch ready for when this change is accepted and merged. Files like BaseAdaptor and FetchFile would require further work.

**Comment by markmcdowall** [comment]
@nerdstrike and @mkszuba see branch [feature/refactor_code_style](https://github.com/Ensembl/ensembl-xref/compare/feature/refactoring_code_style) for initial changes to remove the majority of the small errors.

As mentioned previously BaseAdaptor.pm and FetchFiles.pm have not been touched until other changes come into the master branch. I'm waiting on #7, once merged into this branch then I'll make the relevant pull request.

**Comment by mkszuba** [comment]
I know, that's why I said we should leave it aside :-) I simply meant that it will take time before the benefits of automating this will become visible. At least as you have pointed out, we haven't got too much non-compliant code in this particular repository.

**Comment by premanand17** [comment]
Agree. Not sure about the POD. Most of the parsers have just one run method. So, do we have to put the description at the class level as in 

https://github.com/Ensembl/ensembl-xref/blob/master/modules/Bio/EnsEMBL/Xref/Parser/XenopusJamboreeParser.pm

or the POD should go to the run method?!

**Comment by markmcdowall** [comment]
This probably needs agreement and merge in soon especially if new parsers are going to be added in from here on in. It is best to capture the errors sooner rather than later.

**Comment by markmcdowall** [comment]
@magaliruffier This is the reason for the feature/refactor_coding_style to get rid of the low hanging fruit and the feature/refactor_BaseAdaptor branch for work relating to changes to the BaseAdaptor for work on refactoring the code.

**Comment by magaliruffier** [comment]
OK, so we should try and integrate the other two PRs before this one. I'll make sure to prod again by the end of the week

**Comment by markmcdowall** [comment]
These are tests that should pass for the new parsers that are getting merged in as well.

**Comment by markmcdowall** [comment]
Once merged if we need to exclude the Schema subsections then the following can be used as a stop-gap:

```
all_critic_ok(
    'modules/Bio/EnsEMBL/Xref/DB.pm',
    'modules/Bio/EnsEMBL/Xref/FetchFiles.pm',
    'modules/Bio/EnsEMBL/Xref/Parser.pm',
    'modules/Bio/EnsEMBL/Xref/Schema.pm',
    'modules/Bio/EnsEMBL/Xref/DBSQL',
    'modules/Bio/EnsEMBL/Xref/Parser',
    'modules/Bio/EnsEMBL/Xref/Test');
```

File: modules/t/pod_coverage.t [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+use strict;
+use warnings;
+
+use Test::More;
+
+my $package_available = eval { use Test::Pod::Coverage };
+plan skip_all => 'Test::Pod::Coverage required' if $package_available;
+
+foreach my $mod ( all_modules('modules') ) {
```
[the remark this answers is not in the document]
Comment (markmcdowall) [reply modules/t/pod_coverage.t]: As we were saying this is working on Travis and works locally following the same procedures as used by travis for running the tests with the parameter set to `'modules'`.

If we want to use the files in the `modules/t` directory and run tests from there then the `travis.yml` file should be updated to match that as the standard operating procedure.

**Pull request merged**